sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala

```diff
@@ -170,6 +170,13 @@ case class InsertIntoHiveTable(
     val partitionColumns = fileSinkConf.getTableInfo.getProperties.getProperty("partition_columns")
     val partitionColumnNames = Option(partitionColumns).map(_.split("/")).getOrElse(Array.empty)
 
+    // All column names in the format of "<column name 1>,<column name 2>,..."
+    val columnsCnt = tableDesc.getProperties.getProperty("columns").split(",").size
+
+    if (columnsCnt + numDynamicPartitions != child.output.size) {
+      throw new SparkException(ErrorMsg.TARGET_TABLE_COLUMN_MISMATCH.getMsg)
```

Comment:
- This check only catches cases when number of columns are different.  The error message would not be precise as it would say `column number/types are different`. 

- You could also add other info to the exception: table names and column counts